In [1]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
train = pd.read_csv('../Data/PJT002_train.csv')
test = pd.read_csv('../Data/PJT002_test.csv')
valid = pd.read_csv('../Data/PJT002_validation.csv')

print('train', train.shape)
print('test', test.shape)
print('valid', valid.shape)

train.sample()

train (59199, 172)
test (2957, 171)
valid (6898, 172)


id             dt_of_fr bldng_us bldng_archtctr  bldng_cnt  \
48704  102821  2014-08-25 00:52:55      NaN            NaN          5   

       bldng_ar  ttl_ar  lnd_ar dt_of_athrztn  ttl_grnd_flr  ...  mlt_us_yn  \
48704    196.56  196.56  4116.0           NaN           NaN  ...          N   

      cctv_dstnc  cctv_in_100m  fr_wthr_fclt_in_100m  tbc_rtl_str_dstnc  \
48704        158             0                     0                266   

       sft_emrgnc_bll_dstnc  ahsm_dstnc  no_tbc_zn_dstnc  bldng_cnt_in_50m  \
48704                 32333       49550               16                 0   

       fr_yn  
48704      N  

[1 rows x 172 columns]

In [4]:
# train-test-valid 데이터셋 모두 합쳐서 전처리
train['tag'] = 'train'
test['tag'] = 'test'
valid['tag'] = 'valid'

df = pd.concat([train, test, valid])
df.sample()

ahsm_dstnc  bldng_ar  bldng_ar_prc bldng_archtctr  bldng_cnt  \
40112       22386       0.0           NaN            NaN          1   

       bldng_cnt_in_50m bldng_us bldng_us_clssfctn  cctv_dstnc  cctv_in_100m  \
40112                 0      NaN               NaN          85             0   

       ... rgnl_ar_nm2 sft_emrgnc_bll_dstnc    tag  tbc_rtl_str_dstnc  tmprtr  \
40112  ...      지정되지않음                 1937  train               1152    28.5   

       ttl_ar  ttl_dwn_flr  ttl_grnd_flr  wnd_drctn  wnd_spd  
40112     0.0          NaN           NaN      230.0      2.2  

[1 rows x 173 columns]

In [5]:
# 컬럼명 확인
pd.DataFrame(df.columns)

0
0              ahsm_dstnc
1                bldng_ar
2            bldng_ar_prc
3          bldng_archtctr
4               bldng_cnt
5        bldng_cnt_in_50m
6                bldng_us
7       bldng_us_clssfctn
8              cctv_dstnc
9            cctv_in_100m
10          dt_of_athrztn
11               dt_of_fr
12    ele_engry_us_201401
13    ele_engry_us_201402
14    ele_engry_us_201403
15    ele_engry_us_201404
16    ele_engry_us_201405
17    ele_engry_us_201406
18    ele_engry_us_201407
19    ele_engry_us_201408
20    ele_engry_us_201409
21    ele_engry_us_201410
22    ele_engry_us_201411
23    ele_engry_us_201412
24    ele_engry_us_201501
25    ele_engry_us_201502
26    ele_engry_us_201503
27    ele_engry_us_201504
28    ele_engry_us_201505
29    ele_engry_us_201506
..                    ...
143        lnd_us_sttn_nm
144           lw_13101010
145           lw_13101110
146           lw_13101210
147           lw_13101211
148           lw_13101310
149           lw_13101410
150           lw_13111010
151           lw_13111110
152           lw_13121010
153           lw_13121011
154           lw_13131010
155           lw_13131110
156           lw_13141010
157           lw_13141011
158             mlt_us_yn
159       no_tbc_zn_dstnc
160               prcpttn
161              rd_sd_nm
162            rgnl_ar_nm
163           rgnl_ar_nm2
164  sft_emrgnc_bll_dstnc
165                   tag
166     tbc_rtl_str_dstnc
167                tmprtr
168                ttl_ar
169           ttl_dwn_flr
170          ttl_grnd_flr
171             wnd_drctn
172               wnd_spd

[173 rows x 1 columns]

## Null 값 확인

In [6]:
train.isnull().sum()

id                          0
dt_of_fr                    0
bldng_us                27677
bldng_archtctr          27665
bldng_cnt                   0
bldng_ar                    0
ttl_ar                      0
lnd_ar                      0
dt_of_athrztn           27581
ttl_grnd_flr            10210
ttl_dwn_flr             11005
bldng_us_clssfctn       29343
tmprtr                     13
prcpttn                 53629
wnd_spd                    33
wnd_drctn                 221
hmdt                       22
gas_engry_us_201401     53825
ele_engry_us_201401     43550
gas_engry_us_201402     53825
ele_engry_us_201402     43550
gas_engry_us_201403     53825
ele_engry_us_201403     43550
gas_engry_us_201404     53825
ele_engry_us_201404     43550
gas_engry_us_201405     53825
ele_engry_us_201405     43550
gas_engry_us_201406     53825
ele_engry_us_201406     43550
gas_engry_us_201407     53825
                        ...  
lw_13111010             58899
lw_13111110             58918
lw_1312101

- ['bldng_us_clssfctn', 29343]건물용도분류명(주거용)   
- ['bldng_us', 27677]건물용도(단독주택)  
- ['bldng_archtctr', 27665]건물구조(벽돌구조)  
- ['dt_of_athrztn', 27581]건물승인일자(19850417)  
- ['bldng_ar_prc', 21895]단위 면적당 건물 가격_2019년 기준(312411)  
- ['ttl_dwn_flr', 11005]건물들의 지하 층수의 합(0)  
- ['ttl_grnd_flr', 10210]건물들의 지상 층수의 합(4)  
- ['lnd_us_sttn_nm', 1776]토지 이용상황명(전)
- ['rd_sd_nm', 1491]도로측면명(맹지)  
- ['rgnl_ar_nm2', 1491]용도지역지구명2(지정되지 않음)  
- ['rgnl_ar_nm', 1491]용도지역지구명(개발제한구역)  
- ['hm_cnt', 701]행정구역 인구(27312)  
- ['wnd_drctn', 221]풍향(270)
- ['fr_mn_cnt', 41]관할 소방서 인원(224)  
- ['wnd_spd', 33]풍속(2.2)
- ['hmdt', 22]습도(77)
- ['tmprtr', 13]온도_c(4.4)  
- ['emd_nm', 4]행정구역명(경상남도 진주시 판문동)  
__________________________________________________
- ['bldng_cnt_in_50m', 0]반경 50M 이내의 건물 수(23)  
- ['no_tbc_zn_dstnc', 0]금연구역과의 최소 거리(235)  
- ['fr_wthr_fclt_in_100m', 0]반경 100M 이내 소방용수 시설 수(2)  
- ['ahsm_dstnc', 0]자동 심장 충격기와의 최소 거리(234)  
- ['sft_emrgnc_bll_dstnc', 0]안전 비상벨과의 최소 거리(233)  
- ['tbc_rtl_str_dstnc', 0]담배 소매점과의 최소 거리(232)  
- ['cctv_in_100m', 0]반경 100M 이내 공공 CCTV(1)  
- ['mlt_us_yn', 0]다중이용시설 포함여부(Y)  
- ['cctv_dstnc', 0]공공 CCTV와의 최소 거리(232)  
- ['fr_wthr_fclt_dstnc', 0]소방용수시설(소화전 등)과의 최소 거리(223)  
- ['fr_sttn_dstnc', 0]119안전센터와의 거리(3222)  
- ['jmk', 0]지적상 지목(답)  
- [dt_of_fr', 0]화재발생일시(2014-03-18 15:23:18)  
- ['lnd_ar', 0]토지면적(390)  
- ['ttl_ar', 0]건물연면적_건물층별합계전체면적(130.845)  
- ['bldng_ar', 0]건물건축면적(130.845)  
- ['bldng_cnt', 0]건물채수(4)  

## 전기/가스 사용량 컬럼 제외
Null 값도 많고, 컬럼 수가 너무 많아서 제외

In [7]:
columns = ['id', 'dt_of_fr', 'bldng_us', 'bldng_archtctr', 'bldng_cnt'
           , 'bldng_ar', 'ttl_ar', 'lnd_ar', 'dt_of_athrztn', 'ttl_grnd_flr'
           , 'ttl_dwn_flr', 'bldng_us_clssfctn', 'tmprtr', 'prcpttn'
           , 'wnd_spd', 'wnd_drctn', 'hmdt'
           , 'jmk' 
           , 'emd_nm', 'hm_cnt'
           , 'fr_sttn_dstnc', 'fr_wthr_fclt_dstnc'
           , 'mlt_us_yn', 'cctv_dstnc', 'cctv_in_100m', 'fr_mn_cnt'
           , 'fr_wthr_fclt_in_100m', 'tbc_rtl_str_dstnc', 'sft_emrgnc_bll_dstnc'
           , 'ahsm_dstnc', 'no_tbc_zn_dstnc', 'bldng_cnt_in_50m', 'fr_yn'
           , 'tag'
          ]

df = train[columns]
df.sample()

id             dt_of_fr bldng_us bldng_archtctr  bldng_cnt  \
54932  108631  2016-11-21 04:12:04      NaN            NaN          2   

       bldng_ar  ttl_ar  lnd_ar dt_of_athrztn  ttl_grnd_flr  ...  \
54932       0.0     0.0     0.0           NaN           NaN  ...   

       cctv_in_100m fr_mn_cnt  fr_wthr_fclt_in_100m  tbc_rtl_str_dstnc  \
54932             1     112.0                     0                 85   

       sft_emrgnc_bll_dstnc  ahsm_dstnc  no_tbc_zn_dstnc bldng_cnt_in_50m  \
54932                 16984        2142              178               10   

      fr_yn    tag  
54932     N  train  

[1 rows x 34 columns]

# PreProcessing

## One Hot Encoding) fr_yn 화재발생여부
- Y: 1
- N: 0

In [8]:
df['fr_yn'].unique()

array(['N', 'Y'], dtype=object)

In [9]:
df['fire_yes'] = (df['fr_yn'] == 'Y')

df[['fr_yn', 'fire_yes']].sample()

fr_yn  fire_yes
16681     Y      True

## One Hot Encoding) mlt_us_yn 다중이용시설 포함여부

In [10]:
df['mlt_us_yn'].unique()

array(['N', 'Y'], dtype=object)

In [11]:
df['mlt_us_yn(encode)'] = df['mlt_us_yn'] == 'Y'

df[['mlt_us_yn', 'mlt_us_yn(encode)']].sample()

mlt_us_yn  mlt_us_yn(encode)
15267         N              False

In [12]:
data_fire = pd.DataFrame(df.groupby('mlt_us_yn(encode)')['fire_yes'].sum()).reset_index()
data_all = pd.DataFrame(df.groupby('mlt_us_yn(encode)')['id'].count()).reset_index()
data = data_fire.merge(data_all, on='mlt_us_yn(encode)')
data['ratio'] = data['fire_yes'] * 100 / data['id']
data

mlt_us_yn(encode)  fire_yes     id      ratio
0              False    7548.0  59010  12.791052
1               True     109.0    189  57.671958

## datetime Parsing) dt_of_fr 화재발생시점/관찰시점

In [13]:
df['year'] = pd.to_datetime(df['dt_of_fr']).dt.year
df['month'] = pd.to_datetime(df['dt_of_fr']).dt.month
df['day'] = pd.to_datetime(df['dt_of_fr']).dt.day
df['hour'] = pd.to_datetime(df['dt_of_fr']).dt.hour
df['minute'] = pd.to_datetime(df['dt_of_fr']).dt.minute
df['second'] = pd.to_datetime(df['dt_of_fr']).dt.second
df['dayofweek'] = pd.to_datetime(df['dt_of_fr']).dt.day_name()

df['year-month'] = df['dt_of_fr'].apply(lambda x: x[:7])
df['year-month-day'] = df['dt_of_fr'].apply(lambda x: x[:10])

df[['dt_of_fr', 'year-month', 'year-month-day', 'year', 'month', 'day', 'dayofweek', 'hour', 'minute', 'second']].sample()

dt_of_fr year-month year-month-day  year  month  day  \
7586  2015-01-11 11:01:38    2015-01     2015-01-11  2015      1   11   

     dayofweek  hour  minute  second  
7586    Sunday    11       1      38

## Null 값 채우기) emd_nm 행정구역명
전체 데이터 중 행정구역명이 Null 값인 데이터는 4개 뿐이라 삭제

In [14]:
df = df.dropna(subset=['emd_nm'])

## 행정구역명 나눠서 컬럼'address1', 'address2'에 담기

In [15]:
df['address1'] = np.nan
df['address2'] = np.nan

df['address1'] = df['emd_nm'].apply(lambda x: str(x).split()[1][:3])
df['address2'] = df['emd_nm'].apply(lambda x: str(x).split()[2])
df['address'] = df['address1'] + ' ' + df['address2']
# 창원시는 따로
df.loc[df['address1'].apply(lambda x: len(x) > 3), 'address2'] = df['emd_nm'].apply(lambda x: str(x).split()[1][3:])

df[['emd_nm', 'address', 'address1', 'address2']].sample()

emd_nm  address address1 address2
45433  경상남도 거창군 가조면  거창군 가조면      거창군      가조면

## Null 값 채우기) hm_cnt 행정구역 인구 

인구 Null 값인 행정구역을 찾아보니, 몇 개 행정구역에 몰려있음

In [16]:
df[df['hm_cnt'].isnull()]['address'].value_counts()

진주시 상대동      245
진주시 하대동      167
창원시 마산회원구     96
창원시 마산합포구     80
함안군 칠원읍       59
창원시 진해구        9
통영시 산양읍        6
통영시 욕지면        6
사천시 동금동        3
고성군 거류면        2
하동군 진교면        2
하동군 금남면        2
창녕군 대합면        2
거제시 거제면        2
통영시 용남면        2
거제시 장목면        1
고성군 회화면        1
사천시 실안동        1
남해군 고현면        1
창녕군 이방면        1
남해군 이동면        1
남해군 남해읍        1
거제시 둔덕면        1
남해군 창선면        1
거제시 아주동        1
통영시 미수동        1
진주시 상평동        1
양산시 하북면        1
통영시 도산면        1
Name: address, dtype: int64

### 행정구역별로 뜯어보기

#### 진주시 인구 Null 값
> 진주시청 제공, 12월말 기준

진주시 동 | 2014 | 2015 | 2016
-|-|-|-
진주시 상대동| 26_019 | 25_599 | 25_321
진주시 하대동| 27_457 | 26_981 | 27_067


In [17]:
df[df['address'] == '진주시 상대동'].groupby('year')['hm_cnt'].median()

year
2014        NaN
2015        NaN
2016        NaN
2017    22764.0
2018    22296.0
Name: hm_cnt, dtype: float64

In [18]:
df[df['address'] == '진주시 하대동'].groupby('year')['hm_cnt'].median()

year
2014        NaN
2015        NaN
2016        NaN
2017    28687.5
2018    28164.0
Name: hm_cnt, dtype: float64

#### 창원시는 null 값 뿐임. 인구 정보 따로 찾아보기
> 창원시청 제공, 12월말 기준

창원시 구 | 2014 | 2015 | 2016 | 2017 | 2018
-|-|-|-|-|-
창원시 마산합포구 | 184_179 | 183_829 | 183_989 | 181_917 | 182_772
창원시 마산회원구 | 216_749 | 214_264 | 213_163 | 203_926 | 201_565
창원시 진해구 | - | 189_463 | 190_780 | - |196_665

In [19]:
df[df['address'] == '창원시 마산회원구'].groupby('year')['hm_cnt'].median()

year
2014   NaN
2015   NaN
2016   NaN
Name: hm_cnt, dtype: float64

In [20]:
df[df['address'] == '창원시 마산합포구'].groupby('year')['hm_cnt'].median()

year
2014   NaN
2015   NaN
2016   NaN
2017   NaN
Name: hm_cnt, dtype: float64

In [21]:
df[df['address'] == '창원시 진해구'].groupby('year')['hm_cnt'].median()

year
2015   NaN
2016   NaN
2018   NaN
Name: hm_cnt, dtype: float64

#### 함안군 칠원읍

In [22]:
df[df['address'] == '함안군 칠원읍'].groupby('year')['hm_cnt'].median()

year
2014        NaN
2015    21131.0
2016    21022.0
2017    20889.5
2018    20672.0
Name: hm_cnt, dtype: float64

#### 사천시 동금동, 사천시 실안동 동단위 자료 없음
사천시 전체에서 적은 비중이라 삭제

In [23]:
df[df['address'] == '사천시 동금동'].groupby('year')['hm_cnt'].median()

year
2014   NaN
2017   NaN
Name: hm_cnt, dtype: float64

In [24]:
df[df['address'] == '사천시 실안동'].groupby('year')['hm_cnt'].median()

year
2017   NaN
Name: hm_cnt, dtype: float64

In [25]:
df[df['address1'] == '사천시']['address'].value_counts()

사천시 사천읍     380
사천시 사남면     278
사천시 용현면     270
사천시 동서동     266
사천시 서포면     262
사천시 곤양면     243
사천시 남양동     239
사천시 정동면     235
사천시 벌룡동     234
사천시 향촌동     209
사천시 곤명면     205
사천시 선구동     146
사천시 동서금동    138
사천시 축동면     127
사천시 동금동       3
사천시 실안동       1
Name: address, dtype: int64

In [26]:
# id 113578, 118226, 104635
# index 20446 20673 20680
df[df['address'] == '사천시 동금동'] 

# id 92201
# index 20676
# df[df['address'] == '사천시 실안동'] 

id             dt_of_fr   bldng_us bldng_archtctr  bldng_cnt  \
20446  113578  2017-06-26 20:12:00  제1종근린생활시설       철근콘크리트구조          1   
20673  118226  2014-08-08 14:26:00  제1종근린생활시설       철근콘크리트구조          1   
20680  104635  2014-05-22 21:38:00  제1종근린생활시설       철근콘크리트구조          1   

       bldng_ar  ttl_ar  lnd_ar dt_of_athrztn  ttl_grnd_flr  ...  day hour  \
20446    225.21  203.73   675.0   2.00502e+07           1.0  ...   26   20   
20673    225.21  203.73   675.0   2.00502e+07           1.0  ...    8   14   
20680    225.21  203.73   675.0   2.00502e+07           1.0  ...   22   21   

       minute  second  dayofweek  year-month  year-month-day address1  \
20446      12       0     Monday     2017-06      2017-06-26      사천시   
20673      26       0     Friday     2014-08      2014-08-08      사천시   
20680      38       0   Thursday     2014-05      2014-05-22      사천시   

      address2  address  
20446      동금동  사천시 동금동  
20673      동금동  사천시 동금동  
20680      동금동  사천시 동금동  

[3 rows x 48 columns]

In [27]:
# 사천시 실안동, 동금동 row 삭제

df = df.drop([20446, 20673, 20680, 20676])

### 연도별 인구 중간값으로 채우기

In [28]:
df['hm_cnt'] = df.groupby(['year', 'address'])['hm_cnt'].transform(lambda x: x.fillna(x.median()))

In [29]:
# 남은 null 값 확인

df['hm_cnt'].isnull().sum()

649

In [30]:
df[df['hm_cnt'].isnull()].groupby(['address', 'year'])['hm_cnt'].size()

address    year
진주시 상대동    2014    80
           2015    79
           2016    86
진주시 하대동    2014    45
           2015    63
           2016    52
창원시 마산합포구  2014    30
           2015    24
           2016    25
           2017     1
창원시 마산회원구  2014    34
           2015    23
           2016    39
창원시 진해구    2015     2
           2016     3
           2018     4
함안군 칠원읍    2014    59
Name: hm_cnt, dtype: int64

In [31]:
df[df['address'] == '함안군 칠원읍'].groupby('year')['hm_cnt'].median()

year
2014        NaN
2015    21131.0
2016    21022.0
2017    20889.5
2018    20672.0
Name: hm_cnt, dtype: float64

In [32]:
함안군_칠원읍_2014_hm_cnt = round(21131 + (21131 - 20672) / 4)  
함안군_칠원읍_2014_hm_cnt

21246

In [33]:
마산회원구 = {2014: 216_749, 2015: 214_264, 2016: 213_163, 2017: 203_926, 2018: 201_565}
마산합포구 = {2014: 184_179, 2015: 183_829, 2016: 183_989, 2017: 181_917, 2018: 182_772}
진해구 = {2014: np.nan, 2015: 189_463, 2016: 190_780, 2017: np.nan, 2018: 196_665}
상대동 = {2014: 26_019, 2015: 25_599, 2016: 25_321, 2017: np.nan, 2018: np.nan}
하대동 = {2014: 27_457, 2015: 26_981, 2016: 27_067, 2017: np.nan, 2018: np.nan}
칠원읍 = {2014: 함안군_칠원읍_2014_hm_cnt, 2015: np.nan, 2016: np.nan, 2017: np.nan, 2018: np.nan}

hm_cnt = pd.DataFrame(data=마산회원구.values(), index=마산회원구.keys(), columns=['마산회원구'])
hm_cnt['마산합포구'] = 마산합포구.values()
hm_cnt['진해구'] = 진해구.values()
hm_cnt['상대동'] = 상대동.values()
hm_cnt['하대동'] = 하대동.values()
hm_cnt['칠원읍'] = 칠원읍.values()

hm_cnt

마산회원구   마산합포구       진해구      상대동      하대동      칠원읍
2014  216749  184179       NaN  26019.0  27457.0  21246.0
2015  214264  183829  189463.0  25599.0  26981.0      NaN
2016  213163  183989  190780.0  25321.0  27067.0      NaN
2017  203926  181917       NaN      NaN      NaN      NaN
2018  201565  182772  196665.0      NaN      NaN      NaN

In [34]:
# 수기로 입력해준 인구 데이터 바탕으로 Null 값 채우기
for place in hm_cnt.columns:
    for year in hm_cnt.index:
        df.loc[(df['hm_cnt'].isnull()) & (df['address2'] == place), 'hm_cnt'] = hm_cnt[place][year]

In [35]:
# Null 값 잘 채워졌는지 확인

df['hm_cnt'].isnull().sum()

0

## Null 값 채우기) fr_mn_cnt 관할소방서 인원 

In [36]:
df['fr_mn_cnt'] = df.groupby(['year', 'address'])['fr_mn_cnt'].transform(lambda x: x.fillna(x.median()))

In [37]:
# 소방서 인원 Null 값 채워졌는지 확인

df['fr_mn_cnt'].isnull().sum()

0

## Null 값 채우기) wnd_spd 풍속

In [38]:
df['wnd_spd'] = df.groupby(['address', 'year-month-day'])['wnd_spd'].transform(lambda x: x.fillna(x.median()))
df['wnd_spd'].isnull().sum()

30

In [39]:
df['wnd_spd'] = df.groupby(['address', 'year-month'])['wnd_spd'].transform(lambda x: x.fillna(x.median()))
df['wnd_spd'].isnull().sum()

0

## Null 값 채우기) wnd_drctn 풍향
1. 해당 지역, 해당일의 풍향 중간값으로 채우기
2. 1에서 안 채워진 경우, 해당 지역, 해당월의 풍향 중간값으로 채우기
3. 2에서 안 채워진 경우, 해당일의 풍향 중간값으로 채우기

In [40]:
df['wnd_drctn'] = df.groupby(['address', 'year-month-day'])['wnd_drctn'].transform(lambda x: x.fillna(x.median()))
df['wnd_drctn'].isnull().sum()

191

In [41]:
df['wnd_drctn'] = df.groupby(['address', 'year-month'])['wnd_drctn'].transform(lambda x: x.fillna(x.median()))
df['wnd_drctn'].isnull().sum()

10

In [42]:
df['wnd_drctn'] = df.groupby(['year-month-day'])['wnd_drctn'].transform(lambda x: x.fillna(x.median()))
df['wnd_drctn'].isnull().sum()

0

## Null 값 채우기) hmdt 습도
corr 으로 습도와 상관계수 높은 컬럼 찾기

In [43]:
corr = train.corr()

In [44]:
corr.loc[corr['hmdt']<-0.2, 'hmdt']

wnd_spd     -0.402712
wnd_drctn   -0.253173
Name: hmdt, dtype: float64

1. 같은 풍속, 풍향인 날의 습도 중간값으로 채우기
2. 1에서 안 채워진 경우, 풍향이 같은 날의 습도 중간값으로 채우기 
3. 2에서 안 채워진 경우, 해당일의 습도 중간값으로 채우기

In [45]:
df['hmdt'].shape

(59191,)

In [46]:
df.shape

(59191, 48)

In [47]:
df['hmdt'] = df.groupby(['wnd_spd', 'wnd_drctn'])['hmdt'].transform(lambda x: x.fillna(x.median()))
df['hmdt'].isnull().sum()

8

In [48]:
df['hmdt'] = df.groupby(['year-month-day'])['hmdt'].transform(lambda x: x.fillna(x.median()))
df['hmdt'].isnull().sum()

0

## Null 값 채우기) tmprtr 온도
corr 으로 온도와 상관계수가 높은 컬럼 찾기

In [49]:
corr.loc[corr['tmprtr'] > 0.1, 'tmprtr']

tmprtr         1.000000
wnd_spd        0.106366
hmdt           0.125568
lw_13101110    0.145238
lw_13101210    0.118775
lw_13101211    0.136644
lw_13111010    0.116014
Name: tmprtr, dtype: float64

1. 풍속과 습도가 같은 날의 온도 중간값으로 채우기
2. 습도가 같은 해당일의 온도 중간값으로 채우기
3. 풍속이 같은 해당일의 온도 중간값으로 채우기
4. 습도가 같은 해당월의 온도 중간값으로 채우기

In [50]:
df.dtypes

id                        int64
dt_of_fr                 object
bldng_us                 object
bldng_archtctr           object
bldng_cnt                 int64
bldng_ar                float64
ttl_ar                  float64
lnd_ar                  float64
dt_of_athrztn            object
ttl_grnd_flr            float64
ttl_dwn_flr             float64
bldng_us_clssfctn        object
tmprtr                  float64
prcpttn                 float64
wnd_spd                 float64
wnd_drctn               float64
hmdt                    float64
jmk                      object
emd_nm                   object
hm_cnt                  float64
fr_sttn_dstnc             int64
fr_wthr_fclt_dstnc        int64
mlt_us_yn                object
cctv_dstnc                int64
cctv_in_100m              int64
fr_mn_cnt               float64
fr_wthr_fclt_in_100m      int64
tbc_rtl_str_dstnc         int64
sft_emrgnc_bll_dstnc      int64
ahsm_dstnc                int64
no_tbc_zn_dstnc           int64
bldng_cn

In [51]:
# wnd_spd 와 hmdt 가 float 라서 같은 값인데 다른 값으로 처리되는 것들이 있음
# 해결하기 위해 int 로 바꾼다

df['wnd_spd(int)'] = df['wnd_spd'].apply(lambda x: int(x))
df['hmdt(int)'] = df['hmdt'].apply(lambda x: int(x))

In [52]:
df['tmprtr'] = df.groupby(['wnd_spd(int)', 'hmdt(int)'])['tmprtr'].transform(lambda x: x.fillna(x.median()))
df['tmprtr'].isnull().sum()

0

# 전처리 완료한 데이터셋 csv 로 저장

In [62]:
# 필요한 컬럼만 불러오기
columns = ['id', 'tag'
           # 날짜
           , 'year', 'month', 'day', 'dayofweek', 'year-month', 'year-month-day', 'hour', 'minute', 'second'
          # 건물 용도
           , 'jmk'
          # 건물 용도 (전처리 안됨)
#            , 'bldng_us', 'bldng_archtctr', 'bldng_us_clssfctn', 'dt_of_athrzt'
          # 건물 정보 (전처리 안됨)
#            , 'bldng_ar', 'ttl_ar', 'lnd_ar', 'ttl_grnd_flr', 'ttl_dwn_flr'
           # 다중이용업소
           , 'mlt_us_yn(encode)'
          # 날씨 정보
           , 'tmprtr', 'wnd_spd', 'wnd_drctn', 'hmdt'
          # 행정구역 정보
           , 'address', 'address1', 'address2', 'hm_cnt'
          # 도심과 가까운 정도
           , 'fr_sttn_dstnc', 'fr_wthr_fclt_dstnc', 'cctv_dstnc', 'cctv_in_100m', 'fr_mn_cnt'
           , 'fr_wthr_fclt_in_100m', 'tbc_rtl_str_dstnc', 'sft_emrgnc_bll_dstnc', 'ahsm_dstnc'
           , 'no_tbc_zn_dstnc', 'bldng_cnt_in_50m'
          # 화재발생 여부
           , 'fire_yes'
          ]

df[columns]

id    tag  year  month  day  dayofweek year-month year-month-day  \
0      133759  train  2015     10    3   Saturday    2015-10     2015-10-03   
1       96248  train  2016     11   13     Sunday    2016-11     2016-11-13   
2       71317  train  2016     11   21     Monday    2016-11     2016-11-21   
3       86638  train  2016     11    7     Monday    2016-11     2016-11-07   
4       97748  train  2014      1   28    Tuesday    2014-01     2014-01-28   
5      124960  train  2016     11    3   Thursday    2016-11     2016-11-03   
6      117245  train  2016     11   26   Saturday    2016-11     2016-11-26   
7      128704  train  2014      1   12     Sunday    2014-01     2014-01-12   
8       69465  train  2014      1   24     Friday    2014-01     2014-01-24   
9       85144  train  2014      1   18   Saturday    2014-01     2014-01-18   
10     131014  train  2014      1   14    Tuesday    2014-01     2014-01-14   
11     131084  train  2016      5   19   Thursday    2016-05     2016-05-19   
12     105894  train  2015      2    9     Monday    2015-02     2015-02-09   
13     122039  train  2015      2   21   Saturday    2015-02     2015-02-21   
14      93985  train  2018      2    6    Tuesday    2018-02     2018-02-06   
15      72445  train  2018      2   11     Sunday    2018-02     2018-02-11   
16      99350  train  2018      2   24   Saturday    2018-02     2018-02-24   
17     102168  train  2018      2   12     Monday    2018-02     2018-02-12   
18     109956  train  2018      2   18     Sunday    2018-02     2018-02-18   
19      89995  train  2016      7   19    Tuesday    2016-07     2016-07-19   
20      87580  train  2016      7   31     Sunday    2016-07     2016-07-31   
21      93856  train  2016      7    4     Monday    2016-07     2016-07-04   
22      97379  train  2016      7   31     Sunday    2016-07     2016-07-31   
23     105441  train  2016      7   10     Sunday    2016-07     2016-07-10   
24     106996  train  2016      7   26    Tuesday    2016-07     2016-07-26   
25     108659  train  2016      7   25     Monday    2016-07     2016-07-25   
26     126168  train  2016      7    7   Thursday    2016-07     2016-07-07   
27     105297  train  2018      4   12   Thursday    2018-04     2018-04-12   
28     110495  train  2018      4   12   Thursday    2018-04     2018-04-12   
29     110806  train  2018      4   29     Sunday    2018-04     2018-04-29   
...       ...    ...   ...    ...  ...        ...        ...            ...   
59169  135254  train  2018     10    3  Wednesday    2018-10     2018-10-03   
59170   88801  train  2017     10   16     Monday    2017-10     2017-10-16   
59171   88934  train  2018     10    5     Friday    2018-10     2018-10-05   
59172  104318  train  2018     10   15     Monday    2018-10     2018-10-15   
59173  104420  train  2018     10   20   Saturday    2018-10     2018-10-20   
59174   97765  train  2015      4   12     Sunday    2015-04     2015-04-12   
59175  106081  train  2015      4   28    Tuesday    2015-04     2015-04-28   
59176  119153  train  2015      4   22  Wednesday    2015-04     2015-04-22   
59177  124407  train  2015      4   26     Sunday    2015-04     2015-04-26   
59178   98654  train  2015      5   13  Wednesday    2015-05     2015-05-13   
59179  119598  train  2015      5   21   Thursday    2015-05     2015-05-21   
59180  117329  train  2016      9   22   Thursday    2016-09     2016-09-22   
59181   98878  train  2018      8   13     Monday    2018-08     2018-08-13   
59182  115627  train  2018      8   24     Friday    2018-08     2018-08-24   
59183   87952  train  2014     10   13     Monday    2014-10     2014-10-13   
59184  111403  train  2014     10   16   Thursday    2014-10     2014-10-16   
59185  107618  train  2017      7   22   Saturday    2017-07     2017-07-22   
59186   82402  train  2017      4   26  Wednesday    2017-04     2017-04-26   
59187  130614  train  2017      4    4    Tuesday    2017-04 

In [67]:
df[columns].to_csv('../Data/preprocessingdf.csv', index=False)

In [68]:
# 잘 저장됐는지 확인
pd.read_csv('../Data/preprocessingdf.csv')

id    tag  year  month  day  dayofweek year-month year-month-day  \
0      133759  train  2015     10    3   Saturday    2015-10     2015-10-03   
1       96248  train  2016     11   13     Sunday    2016-11     2016-11-13   
2       71317  train  2016     11   21     Monday    2016-11     2016-11-21   
3       86638  train  2016     11    7     Monday    2016-11     2016-11-07   
4       97748  train  2014      1   28    Tuesday    2014-01     2014-01-28   
5      124960  train  2016     11    3   Thursday    2016-11     2016-11-03   
6      117245  train  2016     11   26   Saturday    2016-11     2016-11-26   
7      128704  train  2014      1   12     Sunday    2014-01     2014-01-12   
8       69465  train  2014      1   24     Friday    2014-01     2014-01-24   
9       85144  train  2014      1   18   Saturday    2014-01     2014-01-18   
10     131014  train  2014      1   14    Tuesday    2014-01     2014-01-14   
11     131084  train  2016      5   19   Thursday    2016-05     2016-05-19   
12     105894  train  2015      2    9     Monday    2015-02     2015-02-09   
13     122039  train  2015      2   21   Saturday    2015-02     2015-02-21   
14      93985  train  2018      2    6    Tuesday    2018-02     2018-02-06   
15      72445  train  2018      2   11     Sunday    2018-02     2018-02-11   
16      99350  train  2018      2   24   Saturday    2018-02     2018-02-24   
17     102168  train  2018      2   12     Monday    2018-02     2018-02-12   
18     109956  train  2018      2   18     Sunday    2018-02     2018-02-18   
19      89995  train  2016      7   19    Tuesday    2016-07     2016-07-19   
20      87580  train  2016      7   31     Sunday    2016-07     2016-07-31   
21      93856  train  2016      7    4     Monday    2016-07     2016-07-04   
22      97379  train  2016      7   31     Sunday    2016-07     2016-07-31   
23     105441  train  2016      7   10     Sunday    2016-07     2016-07-10   
24     106996  train  2016      7   26    Tuesday    2016-07     2016-07-26   
25     108659  train  2016      7   25     Monday    2016-07     2016-07-25   
26     126168  train  2016      7    7   Thursday    2016-07     2016-07-07   
27     105297  train  2018      4   12   Thursday    2018-04     2018-04-12   
28     110495  train  2018      4   12   Thursday    2018-04     2018-04-12   
29     110806  train  2018      4   29     Sunday    2018-04     2018-04-29   
...       ...    ...   ...    ...  ...        ...        ...            ...   
59161  135254  train  2018     10    3  Wednesday    2018-10     2018-10-03   
59162   88801  train  2017     10   16     Monday    2017-10     2017-10-16   
59163   88934  train  2018     10    5     Friday    2018-10     2018-10-05   
59164  104318  train  2018     10   15     Monday    2018-10     2018-10-15   
59165  104420  train  2018     10   20   Saturday    2018-10     2018-10-20   
59166   97765  train  2015      4   12     Sunday    2015-04     2015-04-12   
59167  106081  train  2015      4   28    Tuesday    2015-04     2015-04-28   
59168  119153  train  2015      4   22  Wednesday    2015-04     2015-04-22   
59169  124407  train  2015      4   26     Sunday    2015-04     2015-04-26   
59170   98654  train  2015      5   13  Wednesday    2015-05     2015-05-13   
59171  119598  train  2015      5   21   Thursday    2015-05     2015-05-21   
59172  117329  train  2016      9   22   Thursday    2016-09     2016-09-22   
59173   98878  train  2018      8   13     Monday    2018-08     2018-08-13   
59174  115627  train  2018      8   24     Friday    2018-08     2018-08-24   
59175   87952  train  2014     10   13     Monday    2014-10     2014-10-13   
59176  111403  train  2014     10   16   Thursday    2014-10     2014-10-16   
59177  107618  train  2017      7   22   Saturday    2017-07     2017-07-22   
59178   82402  train  2017      4   26  Wednesday    2017-04     2017-04-26   
59179  130614  train  2017      4    4    Tuesday    2017-04 